### PR TITLE
Add the initial implementation of the LDAP package

### PR DESCRIPTION
**Pull request by Nuvindu**
$subject

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+      - main
+      - 2201.[0-9]+.x
+  repository_dispatch:
+    types: check_connector_for_breaking_changes
+
+jobs:
+  call_workflow:
+    name: Run Connector Build Workflow
+    if: ${{ github.repository_owner == 'ballerina-platform' }}
+    uses: ballerina-platform/ballerina-library/.github/workflows/build-connector-template.yml@main
```
Comment (NipunaRanasinghe) [inline .github/workflows/ci.yml]: @Nuvindu We should be using stdlib templetes (instead of using connector templates) in our workflows

Comment (Nuvindu) [reply .github/workflows/ci.yml]: Yeah, will fix that.

File: ballerina/client.bal [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    #
+    # + config - The configurations to be used when initializing the client
+    # + return - An error if client initialization failed
+    public isolated function init(*ConnectionConfig config) returns error? {
+        self.generateLdapClient(config);
+    }
+
+    private isolated function generateLdapClient(ConnectionConfig config) = @java:Method {
+        'class: "io.ballerina.lib.ldap.Ldap"
+    } external;
```
Comment (NipunaRanasinghe) [inline ballerina/client.bal]: How about

```suggestion
    private isolated function initLdapConnection(ConnectionConfig config) = @java:Method {
        'class: "io.ballerina.lib.ldap.Ldap"
    } external;
```
?

IMO here we are not "generating" something, rather initialisialising a native level client/connection, right?

File: native/src/main/java/io/ballerina/lib/ldap/Ldap.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        try {
+            LDAPConnection ldapConnection = (LDAPConnection) ldapClient.getNativeData(ModuleUtils.NATIVE_CLIENT);
+            List<Modification> modificationList = new ArrayList<>();
+            for (BString key: entry.getKeys()) {
+                modificationList
+                        .add(new Modification(ModificationType.REPLACE, key.getValue(), entry.get(key).getValue()));
+            }
+            ModifyRequest modifyRequest = new ModifyRequest(distinguishedName.getValue(), modificationList);
+            ldapConnection.modify(modifyRequest);
+            return null;
```
Comment (NipunaRanasinghe) [inline native/src/main/java/io/ballerina/lib/ldap/Ldap.java]: any reason to return null? are we not getting any response from the native library API?

File: native/src/main/java/module-info.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+module io.ballerina.lib.microsoft.ad {
```
Comment (NipunaRanasinghe) [inline native/src/main/java/module-info.java]:
```suggestion
module io.ballerina.lib.ldap {
```

File: ballerina/client.bal (hunk 2 above)
Comment (Nuvindu) [reply ballerina/client.bal]: Yeah. Fixed it

File: native/src/main/java/io/ballerina/lib/ldap/Ldap.java (hunk 3 above)
Comment (Nuvindu) [reply native/src/main/java/io/ballerina/lib/ldap/Ldap.java]: Fixed in [9ed07e5](https://github.com/Nuvindu/module-ldap/pull/1/commits/9ed07e5ae804fe6a0d431e4623e4af37e2e931dc)

File: ballerina/types.bal [hunk 5]
```diff
@@ -50,3 +50,51 @@ public type UserConfig record {
     string distinguishedName?;
     string manager?;
 };
+
+public type LDAPResponse record {
+    string matchedDN;
+    ResultCode resultCode;
+    string operationType;
+};
```
Comment (NipunaRanasinghe) [inline ballerina/types.bal]: I guess we can make this a closed record

File: ballerina/types.bal [hunk 6]
```diff
@@ -50,3 +50,51 @@ public type UserConfig record {
     string distinguishedName?;
     string manager?;
 };
+
```
Comment (NipunaRanasinghe) [inline ballerina/types.bal]: missing docs for public type `LDAPResponse`

File: native/src/main/java/io/ballerina/lib/ldap/Utils.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+package io.ballerina.lib.ldap;
+
+import io.ballerina.runtime.api.creators.ErrorCreator;
+import io.ballerina.runtime.api.values.BError;
+
+import static io.ballerina.lib.ldap.ModuleUtils.getModule;
+import static io.ballerina.runtime.api.utils.StringUtils.fromString;
+
+public final class Utils {
```
Comment (NipunaRanasinghe) [inline native/src/main/java/io/ballerina/lib/ldap/Utils.java:36]: missing java docs. Please update all the newly added classes with java docs

Comment (Nuvindu) [reply native/src/main/java/io/ballerina/lib/ldap/Utils.java:36]: Fixed in [9773db2](https://github.com/Nuvindu/module-ldap/pull/1/commits/9773db2c06375f79092e796c148454d63fa0a527)

**Pull request merged**